plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergStatistics.java

```diff
@@ -0,0 +1,325 @@
+            }
+            else {
+                this.nullCounts.put(id, Optional.empty());
+            }
+        }
+
+        private void updateMinMaxStats(
+                int id,
+                io.trino.spi.type.Type type,
+                Object lowerBound,
+                Object upperBound,
+                Optional<Long> nullCount,
+                long recordCount)
+        {
+            // If this column is only nulls for this file, don't update or invalidate min/max statistics
+            if (nullCount.isEmpty() || nullCount.get() != recordCount) {
+                columnStatistics.putIfAbsent(id, new ColumnStatistics(type, typeManager, lowerBound, upperBound));
+                columnStatistics.get(id).updateMinMax(lowerBound, upperBound);
+            }
+        }
+
+        private static Map<Integer, Object> convertBounds(Map<Integer, Type.PrimitiveType> idToTypeMapping, Map<Integer, ByteBuffer> idToMetricMap)
+        {
+            if (idToMetricMap == null) {
+                return null;
+            }
+            ImmutableMap.Builder<Integer, Object> map = ImmutableMap.builder();
+            idToMetricMap.forEach((id, value) -> {
+                Type.PrimitiveType type = idToTypeMapping.get(id);
+                verify(type != null, "No type for column id %s, known types: %s", id, idToTypeMapping);
+                map.put(id, Conversions.fromByteBuffer(type, value));
+            });
+            return map.build();
+        }
+    }
+
+    private static class ColumnStatistics
+    {
+        private final io.trino.spi.type.Type type;
+        private final TypeManager typeManager;
+        private Optional<Object> min;
+        private Optional<Object> max;
+
+        public ColumnStatistics(io.trino.spi.type.Type type, TypeManager typeManager, Object initialMin, Object initialMax)
+        {
+            this.type = requireNonNull(type, "type is null");
+            this.typeManager = requireNonNull(typeManager, "typeManager is null");
+            this.min = Optional.ofNullable(initialMin);
+            this.max = Optional.ofNullable(initialMax);
+        }
+
+        public Optional<Object> getMin()
+        {
+            return min;
+        }
+
+        public Optional<Object> getMax()
+        {
+            return max;
+        }
+
+        /**
+         * @param min Trino encoded minimum value
+         * @param max Trino encoded maximum value
+         */
+        public void updateMinMax(Object min, Object max)
+        {
+            // Update the stats, as long as they haven't already been invalidated
+            if (this.min.isPresent()) {
+                if (min == null) {
+                    this.min = Optional.empty();
+                }
+                else if (compareTrinoValue(min, this.min.get()) < 0) {
+                    this.min = Optional.of(min);
+                }
+            }
```

Comment:
This could use `ifPresent`, or even better, it could use `map` or `mergeOptionals` (suggested above).

```java
this.min = this.min.map(currentMin ->
        min != null && compareTrinoValue(min, currentMin) < 0 ? newValue : currentMin);

this.min = mergeOptionals(this.min, Optional.ofNullable(min), (currentValue, newValue) ->
        compareTrinoValue(newValue, currentMin) < 0 ? newValue : currentValue);
```

(If you make the bound variables `Optional`s as I suggested in an earlier comment, then I think `mergeOptionals` is best here. Otherwise, I think the `map` version is better.)

Reply:
These miss a case where we want to invalidate the stats. For example, if the first file has stats for a column and then the second file does not, we should treat that the same as if the order is reversed.